gammapy/utils/random/inverse_cdf.py

```diff
@@ -77,3 +83,125 @@ def sample(self, size):
 
         index = index + self.random_state.uniform(low=-0.5, high=0.5, size=index.shape)
         return index
+
+
+class MapEventSampler:
+    """Map event sampler.
+
+    Parameters
+    ----------
+
+    npred_map : `~gammapy.maps.Map`
+            Predicted number of counts map.
+    random_state : {int, 'random-seed', 'global-rng', `~numpy.random.RandomState`}
+            Defines random number generator initialisation.
+            Passed to `~gammapy.utils.random.get_random_state`.
+    lc : `~gammapy.time.models.LightCurveTableModel`
+            Input light-curve model of the source, given with columns labelled
+            as "time" (in seconds) and "normalization" (arbitrary units): the bin time
+            HAS to be costant.
+    phase_lc : `~gammapy.time.models.PhaseCurveTableModel`
+            Input phase-curve model of the source, given with columns labelled
+            as "Phase" and "normalization" (arbitrary units): the bin time
+            HAS to be costant.
+    tmin : float
+            Start time of the sampling, given in seconds.
+    tmax : float
+            Stop time of the sampling, given in seconds.
+    """
+
+    def __init__(
+        self, npred_map, random_state=0, lc=None, phase_lc=None, tmin=0, tmax=3600
+    ):
+        self.random_state = get_random_state(random_state)
+        self.npred_map = npred_map
+        self.lc = lc
+        self.phase_lc = phase_lc
+        self.tmin = tmin
+        self.tmax = tmax
+
+    def npred_total(self):
+        """ Calculate the total number of the sampled predicted events.
+            
+        Returns
+        -------
+        random_state.poisson() : int
+                        Number of predicted events.
+        """
+
+        return self.random_state.poisson(np.sum(self.npred_map.data))
+
+    def sample_npred(self):
+        """ Calculate energy and Galactic coordinates of the sampled source events.
+            
+        Returns
+        -------
+        coords : `~numpy.ndarray`
+                Array with coordinates and energies of the sampled events.
+        """
+
+        self.n_events = self.npred_total()
```

Comment:
Once `npred_total` is a lazyproperty you can just access `self.npred_total` and don't have to set `self.n_events`.

Reply:
Done.